### DISCUSSION TITLE
Flask SocketIO Server Emit works intermittently

### DISCUSSION THREAD

**Discussion by csharmatech** [miguelgrinberg/Flask-SocketIO#1917] [Q&A]
Hi Miguel,
I am using Flask SocketIO with Nginx, gunicorn for catching the client events(triggered by button click) on the server side which does some processing and then emits the response back to client asynchronously. For sending asynchronous updates from server, I am using threading such that long running task (which is run in a thread) is executed and once completes sends it to client. The problem I am facing is that server emit works intermittently. Not sure where the problem lies? In the logs, I do see the server emit data.

**Client side Javascript**
```
    <script>
        var socket = io();
        $(document).ready(function() {
        var socket = io.connect('http://flask:5000');
        });

function myupdate() {
        socket.emit("my_event", function(msg) {
       });
}
socket.on("final", function(msg) {
document.getElementById("two").innerHTML=msg.data1;
document.getElementById("four").innerHTML=msg.data2;
});
</script>
<body>
<button onclick="myupdate();">Websocket Asynchronous update</button>
<pre id="place1"></pre>
<pre id="place2"></pre>
</body>
```

**Server Side Python Code**

```
def longdata():
    listing1 = subprocess.run(['ping','-c3', 'google.com'],stdout=subprocess.PIPE,universal_newlines=True)
    longrunning = listing1.stdout
    return longrunning

@app.route('/')
def index():
    return render_template('base.html')

@socketio.on('my_event')
def handle_my_custom_event():
        mythread = threading.Thread(target=longdata) 
        mythread.start()
        while mythread.is_alive():
            emit('final', {'data1': "Getting ping results....", 'data2': "Hold on!!\n"})
            if not mythread.is_alive():
                emit('final', {'data1': longdata(), 'data2': "Here are the results"})
                break;

```

**Comment by csharmatech** [comment]
I am using gevent as below:

`gunicorn -k geventwebsocket.gunicorn.workers.GeventWebSocketWorker -w 1 myapp:app --bind 0.0.0.0:5000`

**Reply by miguelgrinberg** [reply]
It would be useful to see logs from the client and server sides for a session, so that we can see where events are getting lost. One possible source of problems is that you are mixing gevent with subprocess. Not sure how reliable that is. I assume you are aware that when using gevent all the blocking functions in the Python standard library are replaced with gevent's own implementations. This is the so called monkey patching, that I believe gunicorn should be doing for you.

**Reply by csharmatech** [reply]
Thanks Miguel for your feedback.

I tried to start the gunicorn using threaded worker as below (instead of using gevent) and it seems to work consistently now emitting the results from server.I tested using 15 attempts and have 100% success rate. When using gevent with same number of attempts, the success rate is 66%.

`gunicorn -w 1 --threads 100 myapp:app --bind 0.0.0.0:5000`

I will try to use monkey patched gevent to see how it performs.

**Reply by miguelgrinberg** [reply]
Oh interesting, so then it seems I was right with my guess. Note that using gevent without monkey patching cannot work, you cannot mix regular threads with gevent greenlets, its either one or the other. The monkey patching will convert your long running thread to a greenlet. As far as I know, gunicorn automatically does this monkey patching.
